### DISCUSSION TITLE
Migrations and Foreign Keys

### DISCUSSION THREAD

**Discussion by iroddis** [SeaQL/sea-orm#909] [Q&A] [answered]
Hello. I'm starting a new project, and using migrations to define all the tables. The documentation isn't completely clear, but I think the process is:

1. Write initial migations
2. Run the migrations against a specific database
3. Generate entities from what's in the database

These are my questions:

- Is there a way to go straight from migrations -> entities, without using the database as a stepping stone?
- If not, is it possible to work with multiple back ends (e.g. `sqlite::memory:` and `posgres`)? Will the entity code be the same?
- Is it possible to use foreign key constraints in migrations without having generated the entities? I normally do one migration-per-table, but table enums are private within migrations, so it's not possible to reference them from other migrations.

**Comment by billy1624** [comment, marked as answer]
Hey @iroddis, welcome!

> 1. Write initial migations
> 2. Run the migrations against a specific database
> 3. Generate entities from what's in the database

This is correct!

> - Is there a way to go straight from migrations -> entities, without using the database as a stepping stone?

Then, you'd need to define the entity yourself manually. So the answer is no.

> - If not, is it possible to work with multiple back ends (e.g. `sqlite::memory:` and `posgres`)? Will the entity code be the same?

It's possible, but I would suggest you to generate the entity from `postgres` / `mysql`. Because `sqlite` doesn't support datetime and the resulting entity will have datetime field defined in `String` instead of `DateTime`

> - Is it possible to use foreign key constraints in migrations without having generated the entities? I normally do one migration-per-table, but table enums are private within migrations, so it's not possible to reference them from other migrations.

You could move the iden enum into a separate module and then public it (i.e. `pub enum ...`). This way, the iden enum could be shared across multiple migrations.

**Reply by iroddis** [reply]
Thank you very much, this is really helpful. I ended up defining all the `Iden` enums in `migrations/src/tables.rs`, and pulling them in from there.

It's working pretty well, although any columns dropped by later migrations need to stay in the enum. The enums will end up a little crufty, as they'll need to have every column ever defined on the table. The entities will be clean, though.
